src/App.tsx

```diff
@@ -1,34 +1,100 @@
-/* eslint-disable max-len */
-import React from 'react';
+import React, { useEffect, useMemo, useState } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+import { Completed, Filters } from './types/Filters';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [filters, setFilters] = useState<Filters>({
+    completedType: Completed.All,
+    searchByText: '',
+  });
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+
+  const selectTodo = (todo: Todo | null) => {
+    setSelectedTodo(todo);
+  };
+
+  const updateFilters = (key: keyof Filters, value: string | Completed) => {
+    setFilters(prevFilters => ({
+      ...prevFilters,
+      [key]: value,
+    }));
+  };
+
+  const filteredTodos = useMemo(() => {
+    return todos.filter(item => {
+      const matchesCompletedType =
+        filters.completedType === Completed.All ||
+        (filters.completedType === Completed.Completed && item.completed) ||
+        (filters.completedType === Completed.Active && !item.completed);
+
+      const matchesSearchText = item.title
+        .toLowerCase()
+        .includes(filters.searchByText.toLowerCase());
+
+      return matchesCompletedType && matchesSearchText;
+    });
+  }, [todos, filters]);
+
+  useEffect(() => {
+    getTodos()
+      .then(todosData => {
+        setTodos(todosData);
+      })
+      .catch(error => {
+        setErrorMessage(`Failed to fetch todos: ${error.message}`);
+      });
+  }, []);
+
   return (
     <>
       <div className="section">
         <div className="container">
           <div className="box">
             <h1 className="title">Todos:</h1>
 
+            {errorMessage && (
+              <p className="notification is-danger">{errorMessage}</p>
+            )}
+
             <div className="block">
-              <TodoFilter />
+              <TodoFilter
+                filterOptions={filters}
+                onFilterChange={updateFilters}
+              />
             </div>
 
             <div className="block">
-              <Loader />
-              <TodoList />
+              {todos.length === 0 ? (
+                <Loader />
```

Comment:
The condition `todos.length === 0` is used to determine when to show the `Loader`. However, this will show the loader even if the todos have been fetched but the list is empty. Consider using a separate state to track the loading status, such as `isLoading`, to ensure the loader is only shown while data is being fetched.

Reply:
> The condition `todos.length === 0` is used to determine when to show the `Loader`. However, this will show the loader even if the todos have been fetched but the list is empty. Consider using a separate state to track the loading status, such as `isLoading`, to ensure the loader is only shown while data is being fetched.

Good point
If you get an empty array from the API, a user will see an infinite loader. You should show the loader only when you actually load something